### ISSUE TITLE
[R-Forge #2548] Add "key" argument to fread()

### ISSUE THREAD

**Issue by arunsrinivasan** [Rdatatable/data.table#590]
Submitted by: Matt Dowle; Assigned to: Nobody; [R-Forge link](https://r-forge.r-project.org/tracker/index.php?func=detail&aid=2548&group_id=240&atid=978)

To make the syntax easier (also if the expected key columns aren't present it can return earlier before starting to read the data) :

http://stackoverflow.com/a/14952167/403310

**Comment by MichaelChirico** [comment]
I want this! Paralelling data.table(...,key="key") should give cleaner code than setkey(fread(...),key)

**Comment by jangorecki** [comment]
@MichaelChirico If you use chaining you can also use `fread(...)[, .SD,, "key"]` which makes it a start point of chain for operations on sorted data.

**Comment by Chingfu** [comment]
The problem still exists as shown below that the values of D60 are all NA for some codes such as those for the code=="6235".

> a <- subset(D, code=="6235")  
> a
      code close         AD   W    D        D3       D3CH D3CH1       D6      D6CH D6CH1      D18      D18CH D18CH1 D60
   1: 6235  9.80 2015-05-15   1    1  9.176667  0.5966667     1 8.338333 0.5416667     1 7.132222 0.18888889      1  NA
   2: 6235  9.16 2015-05-14   1    2  8.580000  0.5566667     1 7.796667 0.4416667     1 6.943333 0.14833333      1  NA
   3: 6235  8.57 2015-05-13   1    3  8.023333  0.5233333     1 7.355000 0.3500000     1 6.795000 0.12055556      1  NA
   4: 6235  8.01 2015-05-12   1    4  7.500000  0.4866667     1 7.005000 0.2433333     1 6.674444 0.09277778      1  NA
   5: 6235  7.49 2015-05-11   1    5  7.013333  0.3266667     1 6.761667 0.1516667     1 6.581667 0.06333333      1  NA
  ---                                                                                                                  
1310: 6235 10.80 2010-01-07 275 1323 10.950000 -0.1500000     0       NA        NA    NA       NA         NA     NA  NA
1311: 6235 11.05 2010-01-06 275 1324 11.100000 -1.1166667     0       NA        NA    NA       NA         NA     NA  NA
1312: 6235 11.00 2010-01-05 275 1325 12.216667         NA    NA       NA        NA    NA       NA         NA     NA  NA
1313: 6235 11.25 2010-01-04 275 1326        NA         NA    NA       NA        NA    NA       NA         NA     NA  NA
1314: 6235 14.40       <NA> 276 1327        NA         NA    NA       NA        NA    NA       NA         NA     NA  NA

-----Original message-----

**Comment by DavidArenburg** [comment]
@Chingfu, how is that related here?

**Comment by Chingfu** [comment]
The problem was solved as follows. 

1. Originally, the following statement got error message "Error: k <= n is not TRUE":
   D[, D60:=            rollapply(close, 60, mean, align='left', fill = NA), by=.(code)]
   
2. Adding na.rm=TRUE in the following statement,  
   D[, D60:=            rollapply(close, 60, mean, align='left', fill = NA, na.rm = TRUE), by=.(code)], 
   I got the message below.
   Error in `[.data.table`(D, , `:=`(D60, rollapply(close, 60, mean, align = "left",  : 
   Type of RHS ('logical') must match LHS ('double'). To check and coerce would impact performance too much for the fastest cases. Either change the type of the target column, or coerce the RHS of := yourself (e.g. by using 1L instead of 1)

3. Adding as.numeric() in the following statement, the error message disappeared:
   D[, D60:= as.numeric(rollapply(close, 60, mean, align='left', fill = NA, na.rm = TRUE)),by=.(code)]
